test/metabase/driver/mysql_test.clj

```diff
@@ -815,3 +815,191 @@
+                                                                                    (lib.metadata/field mp (mt/id :people :zip))] nil)
+                                                :db-type "CHAR"}]]]
+                {:keys [expression db-type]} expressions]
+          (let [nested-query (-> (lib/query mp (lib.metadata/table mp (mt/id table)))
+                                 (lib/with-fields [])
+                                 (lib/expression "UNCASTED" expression)
+                                 (lib/limit 10))]
+            (testing (str "Casting " db-type " to integer")
+              (mt/with-temp
+                [:model/Card
+                 {card-id :id}
+                 {:dataset_query nested-query
+                  :result_metadata (-> (qp/process-query nested-query) :data :results_metadata :columns)
+                  :type :question}]
+                (let [query (-> (lib/query mp (lib.metadata/card mp card-id))
+                                (as-> q
+                                      (lib/expression q "INTCAST" (lib/expression-clause :integer [(->> q lib/visible-columns (filter #(= "UNCASTED" (:name %))) first)] nil)))
+                                (lib/limit 10))
+                      result (-> query (check-integer-query db-type "`subquery`.`UNCASTED`") qp/process-query)
+                      cols (mt/cols result)
+                      rows (mt/rows result)]
+                  (is (= :type/BigInteger (-> cols first :base_type)))
+                  (doseq [[casted-value equals? uncasted-value] rows]
+                    (is (int? casted-value))
+                    (is equals? (str "Not equal for: " casted-value " " uncasted-value))))))))))))
+
+(deftest ^:parallel integer-cast-nested-custom-expressions
+  (mt/test-driver :mysql
+    (mt/dataset test-data
+      (let [mp (mt/metadata-provider)]
+        (doseq [[table expressions] [[:people [{:expression [(lib/expression-clause :concat
+                                                                                    [(lib.metadata/field mp (mt/id :people :id))
+                                                                                     (lib.metadata/field mp (mt/id :people :zip))] nil)
+                                                             (lib/expression-clause :concat
+                                                                                    [(lib.metadata/field mp (mt/id :people :id))
+                                                                                     (lib.metadata/field mp (mt/id :people :zip))] nil)]
+                                                :db-type "CHAR"}]]]
+                {db-type :db-type [e1 e2] :expression} expressions]
+          (let [query (-> (lib/query mp (lib.metadata/table mp (mt/id table)))
+                          (lib/expression "UNCASTED" e1)
+                          (lib/expression "INTCAST" (lib/expression-clause :integer [e2] nil))
```

Comment:
Likewise, this should be `(lib/integer e2)`, calling `lib.expression/integer`.

Reply:
👍